swarm/src/lib.rs

```diff
@@ -1132,39 +1134,69 @@ where
 
                 self.pending_handler_event = Some((peer_id, handler, event));
             }
-            ToSwarm::NewExternalAddrCandidate(addr) => {
+            ToSwarm::NewExternalAddrCandidate {
+                endpoint,
```

Comment:
I am not a fan of this API change. If we don't actually need any state from the TCP transport to do the translation, perhaps this is the time to just remove `Transport::address_translation` entirely?

AFAIK, all transports just delegate to the generic `address_translation` function anyway! Plus, all behaviours get told about all listen addresses. So, can we instead:

- Remember all listen addresses in the identify behaviour
- Call the generic `address_translation` function from there (we might even be able to move it to the identify behaviour which would have the added benefit of it being removed from the public API of `libp2p-core`!)
- In the identify behaviour, we know which addresses need translation, so we can call it only for those
- All other addresses we can just emit as is

Reply:
Thanks for your advise Thomas !
I implemented the address translation without using `Transport::address_translation`
(I did not removed yet `Transport::address_translation` but it is not used anymore)

- listen addresses are already stored in identify
- translation is done when both conditions are meet: 
    -  a multiaddr of a transport requiring translation (use `is_addr_tcp` and `is_quic_addr` taken from the tcp/quic transport implementations)
    - connection was dialed without port reuse
- call the `address_translation` function
- If translation gave an address emit the `ToSwarm::NewExternalAddrCandidate` with the translated address,  otherwise emit the `ToSwarm::NewExternalAddrCandidate` with the observed address,

If this approach is ok, i will :
- remove `Transport::address_translation` function
- remove the `is_addr_tcp` and `is_quic_addr` for the transports
- keep the generic `address_translation` in `lib_p2p_core` because used in `mdns` (but not for nat purpose)